File: src/main/java/net/gegy1000/terrarium/server/capability/TerrariumWorldData.java
package net.gegy1000.terrarium.server.capability;

import net.gegy1000.terrarium.server.world.TerrariumDataInitializer;
import net.gegy1000.terrarium.server.world.TerrariumGeneratorInitializer;
import net.gegy1000.terrarium.server.world.TerrariumWorldType;
import net.gegy1000.terrarium.server.world.coordinate.Coordinate;
import net.gegy1000.terrarium.server.world.generator.ChunkCompositionProcedure;
import net.gegy1000.terrarium.server.world.generator.TerrariumGenerator;
import net.gegy1000.terrarium.server.world.generator.customization.GenerationSettings;
import net.gegy1000.terrarium.server.world.pipeline.data.ColumnDataCache;
import net.gegy1000.terrarium.server.world.pipeline.data.ColumnDataGenerator;
import net.gegy1000.terrarium.server.world.pipeline.data.DataEngine;
import net.gegy1000.terrarium.server.world.pipeline.data.OffThreadDataGenerator;
import net.minecraft.util.EnumFacing;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.ICapabilityProvider;

import javax.annotation.Nullable;

public interface TerrariumWorldData extends ICapabilityProvider {
    ThreadLocal<Boolean> PREVIEW_WORLD = ThreadLocal.withInitial(() -> false);

    @Nullable
    static TerrariumWorldData get(World world) {
        return world.getCapability(TerrariumCapabilities.worldDataCapability, null);
    }

    GenerationSettings getSettings();

    ColumnDataCache getDataCache();

    ChunkCompositionProcedure getCompositionProcedure();

    Coordinate getSpawnPosition();

    class Implementation implements TerrariumWorldData {
        private final GenerationSettings settings;
        private final TerrariumGenerator generator;
        private final ColumnDataCache dataCache;

        public Implementation(World world, TerrariumWorldType worldType) {
            this.settings = GenerationSettings.parse(world);

            TerrariumGeneratorInitializer generatorInitializer = worldType.createGeneratorInitializer(world, this.settings);
            TerrariumDataInitializer dataInitializer = worldType.createDataInitializer(world, this.settings);

            this.generator = generatorInitializer.buildGenerator(PREVIEW_WORLD.get());
            DataEngine engine = dataInitializer.buildDataEngine();

            ColumnDataGenerator dataGenerator = new OffThreadDataGenerator(engine);
            this.dataCache = new ColumnDataCache(world, dataGenerator);
        }

        @Override
        public GenerationSettings getSettings() {
            return this.settings;
        }

        @Override
        public ColumnDataCache getDataCache() {
            return this.dataCache;
        }

        @Override
        public ChunkCompositionProcedure getCompositionProcedure() {
            return this.generator.getCompositionProcedure();
        }

        @Override
        public Coordinate getSpawnPosition() {
            return this.generator.getSpawnPosition();
        }

        @Override
        public boolean hasCapability(Capability<?> capability, EnumFacing facing) {
            return capability == TerrariumCapabilities.worldDataCapability;
        }

        @Override
        public <T> T getCapability(Capability<T> capability, EnumFacing facing) {
            if (capability == TerrariumCapabilities.worldDataCapability) {
                return TerrariumCapabilities.worldDataCapability.cast(this);
            }
            return null;
        }
    }
}

File: src/main/kotlin/net/gegy1000/terrarium/server/capability/TerrariumWorldData.kt
package net.gegy1000.terrarium.server.capability

import net.gegy1000.terrarium.server.map.source.GeocodingSource
import net.gegy1000.terrarium.server.map.source.glob.GlobSource
import net.gegy1000.terrarium.server.map.source.height.HeightSource
import net.gegy1000.terrarium.server.map.source.osm.OverpassSource
import net.gegy1000.terrarium.server.map.source.tiled.TiledSource
import net.gegy1000.terrarium.server.world.EarthGenerationSettings
import net.gegy1000.terrarium.server.world.generator.EarthGenerationHandler
import net.minecraft.util.EnumFacing
import net.minecraft.world.World
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.ICapabilityProvider

interface TerrariumWorldData : ICapabilityProvider {
    val generationHandler: EarthGenerationHandler

    val heightSource: HeightSource
    val globSource: GlobSource
    val overpassSource: OverpassSource
    val geocodingSource: GeocodingSource

    val tiledBufferSources: Set<TiledSource<*>>

    class Implementation(world: World) : TerrariumWorldData {
        private val settings = EarthGenerationSettings.deserialize(world.worldInfo.generatorOptions)

        override val generationHandler: EarthGenerationHandler

        override val heightSource = HeightSource(settings)
        override val globSource = GlobSource(settings)
        override val overpassSource = OverpassSource(settings)
        override val geocodingSource = GeocodingSource(settings)

        override val tiledBufferSources: Set<TiledSource<*>> = setOf(heightSource, globSource, overpassSource)

        init {
            generationHandler = EarthGenerationHandler(this, settings, world.height - 1)

            overpassSource.loadQuery()
        }

        override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
            return capability == TerrariumCapabilities.worldDataCapability
        }

        override fun <T> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
            if (this.hasCapability(capability, facing)) {
                return TerrariumCapabilities.worldDataCapability.cast(this)
            }
            return null
        }
    }
}
